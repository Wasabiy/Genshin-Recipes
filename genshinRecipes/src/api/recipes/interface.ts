export interface Food {
    name?: string,
    rarity?: number,
    type?: Type,
    effect?: string,
    hasRecipe?: boolean,
    description?: string,
    character?: string,
    baseDish?: string,
    recipe?: Recipe[],
    proficiency?: number | string,
    event?: string,
    charcter?: string,
    food: any
}

export interface Recipe {
    item:     string;
    quantity: number;
}

export enum Type {
    ATKBoostingDish = "ATK-Boosting Dish",
    AdventurerSDish = "Adventurer's Dish",
    DEFBoostingDish = "DEF-Boosting Dish",
    RecoveryDish = "Recovery Dish",
}