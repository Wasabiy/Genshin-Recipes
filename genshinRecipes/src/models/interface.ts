export type KeyFood = {
    key: string,
} & Food

export interface Food {
    name: string,
    rarity: number,
    type: DishType,
    effect: string,
    hasRecipe?: boolean,
    description: string,
    character?: string,
    baseDish?: string,
    recipe?: Recipe[],
    proficiency: number | string,
    event?: string,
    isFavorited?: boolean,
}

export interface Recipe {
    item:     string;
    quantity: number;
}
export interface Ingredient{
    name: string,
    description: string,
    sources: string[]
}
export type KeyIngredient = {
    key: string,
} & Ingredient

export enum DishType {
    ATKBoostingDish = "ATK-Boosting Dish",
    AdventurerSDish = "Adventurer's Dish",
    DEFBoostingDish = "DEF-Boosting Dish",
    RecoveryDish = "Recovery Dish",
}