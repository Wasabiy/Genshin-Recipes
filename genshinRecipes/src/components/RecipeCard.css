.cardSquares {
  display: flex;
  flex-direction: column;
  background-color: none;
  margin: 3px;
  padding: 5px;
  width: 120px;
  height: 120px;
  font-size-adjust: inherit;
}
.imageButton {
  position: relative;
  margin: 0;
}

.likedButton {
  position: absolute;
  border: none;
  height: 2rem;
  top: 2px;
  right: 2px;
}

.normalButton {
  position: absolute;
  border: none;
  height: 2rem;
  top: 2px;
  right: 2px;
  border-radius: 10px;
}

.normalButton:hover {
  background-color: rosybrown;
}

.likedButton:hover {
  background-color: rosybrown;
  border-radius: 10px;
}

.recipeImages {
  display: flex;
  width: 80px;
  height: 80px;
  margin: 0;
  padding: 0;
}
.recipeNames {
  display: flex;
  margin: 2px;
  font-size: inherit;
}
