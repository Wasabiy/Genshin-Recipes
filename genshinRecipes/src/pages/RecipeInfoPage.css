#likedBtn:hover {
    background-color: rosybrown;
    border-radius: 10px;
}

#notLikedBtn:hover {
    background-color: rosybrown;
    border-radius: 10px;
}


