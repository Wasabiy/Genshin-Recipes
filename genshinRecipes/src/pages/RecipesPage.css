#displayBox {
  text-align: center;
  width: fit-content;
  overflow-y: scroll;
  height: 48rem;
  width: 60vw;
  border: 0.1vh solid white;
}

#recipesBox a {
  text-decoration: none;
  color: white;
}

#recipesBox {
  max-width: 1200px;
  justify-content: space-between;
  align-content: center;
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  padding: 0;
  font-size: small;
}

#filters {
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  width: fit-content;
  padding: 10vh;
}

.filterList {
  display: flex;
  flex-direction: column;
  gap: 10px;
  height: 27vh;
  align-items: center;
  overflow-y: scroll;
}

.filterList button:hover {
  background-color: #ecb864;
}

.filterList label {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border: 1px;
  border-radius: 20px;
}

.filterList label:hover {
  background-color: #f39c12;
}

.filterList label:focus {
  background-color: #f39c12;
}

.filterList input[type='checkbox'] {
  margin-right: 10px;
}

.filterList input[type='checkbox']:checked {
  label {
    background-color: #f39c12;
    color: #f39c12;
  }
}
.filterList input[type='checkbox']:focus {
  label {
    background-color: #f39c12;
  }
}

.filterList img {
  width: 40px;
  height: auto;
  margin-left: 10px;
}

#ingredientScroll {
  overflow-y: scroll;
}

#allRecipeSection {
  display: flex;
  flex-direction: row;
  align-items: center;
}

@media (max-width: 600px) {
  #allRecipeSection {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    margin: auto;
  }

  #recipesBox {
    justify-content: space-between;
    align-content: center;
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;
    padding: 0;
    font-size: small;
  }

  #filters {
    display: flex;
    align-items: flex-start;
    flex-direction: column;
    width: fit-content;
    padding: 7vh;
    align-items: center;
  }

  #filters a {
    padding: 5vh;
  }

  .filterList {
    display: flex;
    flex-direction: column;
    gap: 10px;
    height: 20vh;
    align-items: center;
    overflow-y: scroll;
    text-align: center;
    border: solid white;
    padding: 3vh;
  }
}
