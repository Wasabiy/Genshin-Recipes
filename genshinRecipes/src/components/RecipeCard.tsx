import { useEffect, useState } from 'react';
import '../style/components/RecipeCard.css';
import 'react-router-dom';
import { KeyFood } from '../models/interface.ts';
import {  useNavigate } from 'react-router-dom';
import { changeFavoriteState, getFavoriteState } from '../utils/globalFunctions.ts';
import like from '../assets/like.png';
import liked from '../assets/liked.png';

function RecipeCard({ food, src, onFavoriteChanged }: { food: KeyFood; src: string; onFavoriteChanged?: () => void }) {
  const [isLiked, setIsLiked] = useState(Boolean);
  const navigate = useNavigate();
  useEffect(() => {
    if (getFavoriteState(food)) {
      setIsLiked(true);
    } else {
      setIsLiked(false);
    }
  }, [food]);

  const handleReadMore = () => {
    navigate('/recipeInfo', { state: { food: food } });
  };
  const handleFavoriteChange = () => {
    const likedBool = getFavoriteState(food);
    setIsLiked(!likedBool);
    changeFavoriteState(food);
    if (onFavoriteChanged) {
      onFavoriteChanged();
    }
  };

  return (
    <>

        <figure key={food.key} className="cardSquares">
          <figure className="imageButton">
            <img className="recipeImages" src={src} alt={food.name} width="50" height="60" onClick={handleReadMore}></img>
            <img
              alt={'likedButton'}
              src={isLiked ? liked : like}
              key={food.key + 'Btn'}
              onClick={(e) => {
                e.preventDefault();
                handleFavoriteChange();
              }}
              className={isLiked ? 'likedButton' : 'normalButton'}
            ></img>
          </figure>
          <p className="recipeNames">{food.name}</p>
        </figure>
    </>
  );
}

export default RecipeCard;
