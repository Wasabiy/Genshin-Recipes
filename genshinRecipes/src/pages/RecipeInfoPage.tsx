import { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Food, KeyFood } from '../models/interface.ts';
import { changeFavoriteState, getFavoriteState } from '../utils/globalFunctions.ts';
import rarityStar from '../assets/rarityStar.png';
import '../style/pages/RecipeInfoPage.css';
import like from '../assets/like.png';
import liked from '../assets/liked.png';
import { QueryClient } from '@tanstack/react-query';
import { QueryClientProvider } from '@tanstack/react-query';

const queryClient = new QueryClient();
export default function RecipeInfoPage({food}:{ food: KeyFood }) {
  return (

      <RecipeInfo keyedFood={food} />

  );
}
function RecipeInfo({keyedFood}:{ keyedFood: KeyFood }) {
  
  const [isLiked, setIsLiked] = useState(Boolean);

  useEffect(() => {
    if (getFavoriteState(keyedFood)) {
      setIsLiked(true);
    } else {
      setIsLiked(false);
    }
  }, [keyedFood]);
  const handleFavoriteChange = () => {
    const likedBool = getFavoriteState(keyedFood);
    setIsLiked(!likedBool);
    changeFavoriteState(keyedFood);
  };

  return (
    <>
      <section id="allFavText">
        <h2>{keyedFood?.name}</h2>
        <h3>Press heart to like or dislike!</h3>
        <img alt={"Image of heart"}
          id="heartImage"
          src={isLiked ? liked : like}
          onClick={(e) => {
            e.preventDefault();
            handleFavoriteChange();
          }}
        ></img>
        <h3>Description:</h3>
        <section id="detailList">
          <ul id="keyedFood">
            <li key={'stars'}>
              Rarity:{' '}
              {[...Array(keyedFood?.rarity)].map((value, index, array) => (
                <img key={'star' + index} src={rarityStar} alt="star" width="20" height="20" />
              ))}
            </li>
            <li key={'type'}> Type: {keyedFood?.type} </li>
            <li key={'proficiency'}> Proficiency: {keyedFood?.proficiency} </li>
            <li key={'effect'}> Effect: {keyedFood?.effect} </li>
            <section id="ingredientBox">
              <h4>Ingredients</h4>
              <ul key={"IngredientUL"} id="ingredientList">
                {keyedFood?.recipe == undefined ? (
                  <p>No Recipe</p>
                ) : (
                  keyedFood?.recipe &&
                  keyedFood?.recipe.map((value) => (
                    <li key={value.item}>
                      {value.item} x {value.quantity}
                    </li>
                  ))
                )}
              </ul>
            </section>
            <section id="descBox">
              <h4 key={"descriptionHeader"} id="descHeader">Description</h4>
              <span id="description">{keyedFood?.description}</span>
            </section>
          </ul>
        </section>
        <section id="imgSection">
          <img
            src={`https://genshin.jmp.blue/consumables/food/${keyedFood?.key}`}
            alt={'Image of the dish named: ' + keyedFood?.name}
          />
        </section>
      </section>
    </>
  );
}
