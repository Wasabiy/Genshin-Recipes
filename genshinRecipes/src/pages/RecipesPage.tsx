import { useState, useEffect } from "react";
import "./RecipesPage.css";
import { KeyFood, KeyIngredient } from "../models/interface.ts";
import RecipeCard from "../components/RecipeCard.tsx";
import {
  QueryClient,
  QueryClientProvider,
  useQuery,
} from "@tanstack/react-query";
import { reformatData } from "../utils/globalFunctions.ts";
import { fetchFood, fetchIngredient } from "../utils/apiCalls.ts";

const queryClient = new QueryClient();
export default function RecipePage() {
  return (
    <QueryClientProvider client={queryClient}>
      <RecipeGen />
    </QueryClientProvider>
  );
}

function RecipeGen() {
  const [itemList, setItemList] = useState<KeyFood[] | null>();
  const [food, setFood] = useState<KeyFood[] | null>();
  const [ingredients, setIngredients] = useState<KeyIngredient[] | null>();

  //fetch data fra APIet
  const { data: foodData, status: foodStatus } = useQuery({
    queryKey: ["foodData"],
    queryFn: fetchFood,
  });
  const { data: ingredientData, status: ingredientStatus } = useQuery({
    queryKey: ["ingredientData"],
    queryFn: fetchIngredient,
  });

  useEffect(() => {
    if (foodStatus == "success") {
      const foods: KeyFood[] = reformatData(foodData, "KeyFood");
      setFood(foods);
      setItemList(foods);
    }
  }, [foodStatus]);
  useEffect(() => {
    if (ingredientStatus == "success") {
      const ingredients: KeyIngredient[] = reformatData(
        ingredientData,
        "KeyIngredient"
      );
      setIngredients(ingredients);
    }
  }, [ingredientStatus]);

  /*function filterRecipes(condition: string, itemList: KeyFood[]) {
        //skriv noe her JayanüíÄ
        /*vi tenkte at denne kan v√¶re for b√•de filtreringsmenyen og s√∏kebaren; om
        man skal filtrere p√• "ATK boosting dish" s√• vil condition v√¶re "atk-boosting-dish" f.eks.. 
        men gj√∏r det du vil seff!
    }*/

  /* function sortRecipes() {
         //do ur thing bestieee <3
     }*/

  return (
    <>
      {foodStatus === "error" && <h2>Error fetching data</h2>}
      {foodStatus === "pending" && <h2>Fetching data...</h2>}
      {foodStatus === "success" && (
        <>
          {" "}
          <section id="title">
            <h2 id="header">Recipes</h2>
            <span id="itemAmount">showing {itemList?.length} recipes</span>
          </section>
          <section id="filterSection">
            <span id="type">Type</span>
            <section id="filterList">
              <button className="filterChoices">Attack boosting</button>
              <button className="filterChoices">DEF boosting</button>
              <button className="filterChoices">Revival dish</button>
              <button className="filterChoices">Healing dish</button>
              <button className="filterChoices">Stamina boosting</button>
            </section>
          </section>
          <span>Ingredients</span>
          <section key="displayIn" id="displayIngredientsBox">
            <section id="ingredientsList">
              {ingredients?.map((value) => {
                return (
                  <label key={value.name}>
                    <input type="checkbox" />
                    {value.name}
                    <img
                      src={`https://genshin.jmp.blue/materials/cooking-ingredients/${value.key}`}
                    />
                  </label>
                );
              })}
            </section>
          </section>
          <section key="displayFood" id="displayBox">
            <section key="recipeBox" id="recipesBox">
              {food?.map((value) => {
                return (
                  <RecipeCard
                    food={value}
                    src={`https://genshin.jmp.blue/consumables/food/${value.key}`}
                  />
                );
              })}
            </section>
          </section>
        </>
      )}
    </>
  );
}
