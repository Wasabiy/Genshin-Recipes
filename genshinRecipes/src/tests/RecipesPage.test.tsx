import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';
import { describe, expect, test, beforeAll, vi } from "vitest";
import { server } from "./mockServer.ts"
import { KeyFood, Type } from "../models/interface.ts";
import RecipeCard from "../components/RecipeCard.tsx";
import RecipesPage from "../pages/RecipesPage.tsx";
import RecipeInfoPage from "../pages/RecipeInfoPage.tsx"
import { changeFavoriteState, getFavoriteState } from "../utils/globalFunctions.ts";
import { MemoryRouter, BrowserRouter as Routes, Route } from "react-router-dom";

vi.mock("../utils/globalFunctions", () => ({
    changeFavoriteState: vi.fn(),
    getFavoriteState: vi.fn(),
}));

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("RecipesPage.tsx snapshot test", () => {
    test('matches snapshot', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <RecipesPage />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

/*describe("RecipesPage.tsx sorting function", () => {

})
*/

