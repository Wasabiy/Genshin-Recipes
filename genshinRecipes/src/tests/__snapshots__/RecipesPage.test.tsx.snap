// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecipesPage.tsx snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <h2>
    Fetching data...
  </h2>
</DocumentFragment>
`;
